observability/src/main/java/io/grpc/observability/interceptors/GcpLogHelper.java

```diff
@@ -0,0 +1,430 @@
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setMetadata(pair.proto)
+          .setPayloadSize(pair.size)
+          .setRpcId(rpcId);
+      if (peerAddress != null) {
+        logEntryBuilder.setPeerAddress(socketToProto(peerAddress));
+      }
+      logger.log(Level.FINE, "Writing Response Header to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs the server trailer.
+     */
+    void logTrailer(
+        long seqId,
+        String serviceName,
+        String methodName,
+        Status status,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(status, "status");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_CLIENT,
+          "peerAddress can only be specified for client");
+
+      PayloadBuilder<GrpcLogRecord.Metadata.Builder> pair = createMetadataProto(metadata);
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setEventType(EventType.GRPC_CALL_TRAILER)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setMetadata(pair.proto)
+          .setPayloadSize(pair.size)
+          .setStatusCode(status.getCode().value())
+          .setRpcId(rpcId);
+      String statusDescription = status.getDescription();
+      if (statusDescription != null) {
+        logEntryBuilder.setStatusMessage(statusDescription);
+      }
+      byte[] statusDetailBytes = metadata.get(STATUS_DETAILS_KEY);
+      if (statusDetailBytes != null) {
+        logEntryBuilder.setStatusDetails(ByteString.copyFrom(statusDetailBytes));
+      }
+      if (peerAddress != null) {
+        logEntryBuilder.setPeerAddress(socketToProto(peerAddress));
+      }
+      logger.log(Level.FINE, "Writing Server Trailer to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs the RPC message.
+     */
+    <T> void logRpcMessage(
+        long seqId,
+        String serviceName,
+        String methodName,
+        GrpcLogRecord.EventType eventType,
+        T message,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          eventType == EventType.GRPC_CALL_REQUEST_MESSAGE
+              || eventType == EventType.GRPC_CALL_RESPONSE_MESSAGE,
+          "event type must correspond to client message or server message");
+      checkNotNull(message, "message");
+
+      // TODO(dnvindhya): Convert message to bystestring
```

Comment:
Why is this a TODO if you already have the code as commented lines? What else is needed to uncomment/finalize the code?

Reply:
Logging RPC message still requires for us to specify request marshaller and response marshaller. I will add this later.